challenge_set_a_table/src/challenge_set_a_table_states/manipulate_machine.py

```diff
@@ -0,0 +1,313 @@
+                                   smach.CBState(lock),
+                                   transitions={'locked': 'GRAB_ITEM'})
+
+            # smach.StateMachine.add("ANNOUNCE_ITEM",
+            #                        states.Say(robot, EntityDescriptionDesignator(self.grab_designator,
+            #                                                                      name="current_item_desc"),
+            #                                   block=False),
+            #                        transitions={'spoken': 'GRAB_ITEM'})
+
+            smach.StateMachine.add("GRAB_ITEM",
+                                   states.Grab(robot, self.grab_designator, self.empty_arm_designator),
+                                   transitions={'done': 'UNLOCK_ITEM_SUCCEED',
+                                                'failed': 'UNLOCK_ITEM_FAIL'})
+
+            @smach.cb_interface(outcomes=["unlocked"])
+            def unlock(userdata=None):
+                """ 'Unlocks' a locking designator """
+                self.grab_designator.unlock()
+
+                return "unlocked"
+
+            smach.StateMachine.add("UNLOCK_ITEM_SUCCEED",
+                                   smach.CBState(unlock),
+                                   transitions={'unlocked': 'succeeded'})
+
+            smach.StateMachine.add("UNLOCK_ITEM_FAIL",
+                                   smach.CBState(unlock),
+                                   transitions={'unlocked': 'failed'})
+
+
+class PlaceSingleItem(smach.State):
+    """ Tries to place an object. A 'place' statemachine is constructed dynamically since this makes it easier to
+     build a statemachine (have we succeeded in grasping the objects?)"""
+    def __init__(self, robot, place_designator):
+        """ Constructor
+
+        :param robot: robot object
+        :param place_designator: Designator that resolves to the pose to place at. E.g. an EmptySpotDesignator
+        """
+        smach.State.__init__(self, outcomes=["succeeded", "failed"])
+
+        self._robot = robot
+        self._place_designator = place_designator
+
+    def execute(self, userdata=None):
+
+        arm = None
+        # See if there's an arm holding something
+        for k, v in self._robot.arms.iteritems():
+            if v.occupied_by is not None:
+                arm = v
+                break
+
+        if arm is None:
+            return "failed"
+
+        # Try to place the object
+        item = ds.EdEntityDesignator(robot=self._robot, id=arm.occupied_by.id)
+        arm_designator = ds.ArmDesignator(all_arms={arm.side: arm}, preferred_arm=arm)
+        sm = states.Place(robot=self._robot, item_to_place=item, place_pose=self._place_designator, arm=arm_designator)
+        result = sm.execute()
+
+        # If failed, do handover to human in order to continue
+        if result != "done":
+            sm = states.HandoverToHuman(robot=self._robot, arm_designator=arm_designator)
+            sm.execute()
```

Comment:
Dynamically creating a state machine is not nice

Reply:
@jlunenburg

Reply:
Point taken but takes too much time to implement